retry.go

```diff
@@ -0,0 +1,155 @@
+package soap
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+type retryTransport struct {
+	maxRetries  int
+	next        http.RoundTripper
+	shouldRetry func(error, *http.Request, *http.Response) bool
+}
+
+func (t *retryTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// if body is present, it must be buffered if there is any chance of a retry
+	// since it can only be consumed once.
+	var br *bytes.Reader
+	if req.Body != nil && req.Body != http.NoBody {
+		var buf bytes.Buffer
+		if _, err := io.Copy(&buf, req.Body); err != nil {
+			req.Body.Close()
+			return nil, fmt.Errorf("error buffering body before retry: %w", err)
+		}
+		req.Body.Close()
+		br = bytes.NewReader(buf.Bytes())
+		req.Body = io.NopCloser(br)
+	}
+	var attemptCount int
+	for {
+		res, err := t.next.RoundTrip(req)
+		attemptCount++
+		if attemptCount-1 >= t.maxRetries {
+			return res, err
+		}
```

Comment:
Unused to seeing `attemptCount-1 >= t.maxRetries` instead of `attemptCount >= t.maxRetries`. E.g. moving `attemptCount++` to after the loop is executed like a "traditional post-increment loop". Is this a Go standard?

Reply:
TBH this is a bit of a copy-paste job from a few existing retry libraries out there - it can be simplified and refined - I'll make a note to have a look at it when I get the chance!